### PR TITLE
Add Table performance improvements

### PR DESCRIPTION
**Pull request by Mogball**
Reduces the number of database requests issued during "Add Table". Testing on a local environment with the example datasets gave about a 4.5x speed up.

- Added more performant "check if table exists" and "get dialect" functions
- Removed queries and calls to `session.commit()` from the inside of loops

### CODE REVIEW

File: superset/models/core.py [hunk 1]
```diff
@@ -743,6 +744,16 @@ def get_perm(self):
...
             "[{obj.database_name}].(id:{obj.id})").format(obj=self)
 
+    def has_table(self, table):
+        engine = self.get_sqla_engine()
+        return engine.dialect.has_table(
+            engine, table.table_name, table.schema or None)
+
+    def get_dialect(self):
+        sqla_url = url.make_url(self.sqlalchemy_uri_decrypted)
+        entrypoint = sqla_url._get_entrypoint()
```
Comment (mistercrunch) [inline superset/models/core.py]: why not `return self.get_sqla_engine().dialect` ?

Comment (Mogball) [reply superset/models/core.py]: There's a bit more overhead to it, plus creating an engine just to get the dialect and then ignoring the instance seems wasteful.

These few lines are what `create_engine` does to determine dialect anyway.

Comment (mistercrunch) [reply superset/models/core.py]: I hate calling private functions in external libs though... there's no guarantees as to whether the api may change in minor version changes.

To mitigate the cost we could change the use of `get_sqla_engine` by using a cached `engine` `@property`. I think it's only a matter of adding a decorator.

Comment (mistercrunch) [reply superset/models/core.py]: check out the `@memoize` decorator in `superset.utils`

Comment (mistercrunch) [reply superset/models/core.py]: also for the sake of moving forward I don't mind if you just call `get_sqla_engine` here and save the `sqla_engine` memoized property for another PR

Comment (Mogball) [reply superset/models/core.py]: Sounds good. I was looking for a way to cache the `sqla_engine` because that function does get called frequently.

Comment (Mogball) [reply superset/models/core.py]: Oh, turns out we can just call `sqla_url.get_dialect()`

**Comment by mistercrunch** [comment]
img

**Pull request merged**